cpp/include/raft/core/pinned_container_policy.hpp

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <cstddef>
+#include <raft/core/mdspan_types.hpp>
+#include <raft/core/resources.hpp>
+#ifndef RAFT_DISABLE_CUDA
+#include <thrust/host_vector.h>
+#include <thrust/mr/allocator.h>
+#include <thrust/system/cuda/memory_resource.h>
+#else
+#include <raft/core/detail/fail_container_policy.hpp>
+#endif
+
+namespace raft {
+#ifndef RAFT_DISABLE_CUDA
+
+/**
+ * @brief A thin wrapper over thrust::host_vector for implementing the pinned mdarray container
+ * policy.
+ *
+ */
+template <typename T>
+struct pinned_container {
```

Comment:
Can the pinned container policy not be created by passing the pinned memory allocator as a template arg to the host_container_policy? The host_container_policy has a template argument for the allocator type.

Reply:
I went down that route initially, but because the thrust and rmm allocators behave a bit differently, this gets tricky. This keeps things a bit less hacky for now, and we can use a pinned rmm allocator when it's available.